feat: show commit list

lorenzo-cavazzi opened the pull request:
Adds all the necessary logic to handle projects commits.

It includes a `ProjectCoordinator` component to handle storing project data to the global Redux store (#623). This created extra difficulties since we can't easily migrate all the `ProjectModel` logic at once. To simplify the temporary co-existence, I created a higher order function `withProjectMapped` to map project data from the global store to single components whenever needed.

The first commit closes #852 by adding a new "Commits" section in the project overview tab, while the second commit re-uses the `CommitsView` component in the merge request section.
The third commit injects all the commits-related logic in the environments components. We may want to clean up a little bit `Notebooks.container.js` when `branches` will be managed by `ProjectCoordinator`.

***Preview***: https://lorenzotest.dev.renku.ch
***How to test***: browse to any project, go to the new "Commits" section on the "Overview" page. Please check also the "Merge request" section and check a notebook in the "File" tab, starting also a new environment from there. This should cover all the relevant changes.

![Screenshot_20200424_194059](https://user-images.githubusercontent.com/43481553/80241253-8e916900-8663-11ea-9f37-800576d0d583.png)

rokroskar commented:
Is there an issue/epic for this?

lorenzo-cavazzi commented:
@rokroskar yes, it's #852 . I accidentally wrote twice #623 in the issue description (fixed) but it was in the `linked issues`

lorenzo-cavazzi commented:
@ciyer good point! We may want to handle pagination better in the future, but in the meanwhile that message would make this limitation clear.
I added the info inside a `<InfoAlert>` Component, let me know if you prefer the plain text.

Here is a comparison (first is the current implementation)

![Screenshot_20200428_190617](https://user-images.githubusercontent.com/43481553/80517130-dd514280-8984-11ea-8f4f-a456c8f4301e.png)

![Screenshot_20200428_190647](https://user-images.githubusercontent.com/43481553/80517141-e0e4c980-8984-11ea-9b84-c2ba452ba59f.png)

lorenzo-cavazzi commented:
@vfried thanks for the suggestions!

I think the current spacing on the top is in line with the other tabs (see Description and Stats). Furthermore, making the top padding `0` would make the top border look thicker (it's visible in your image).

Im not sure about moving the buttons to the right. I see that it would make the page more similar to the `Cards` in the Files tab, but it would be awkward to have the number and the refresh button far from the `Commits` text they qualify.

ciyer commented:
> @ciyer good point! We may want to handle pagination better in the future, but in the meanwhile that message would make this limitation clear.
> I added the info inside a `<InfoAlert>` Component, let me know if you prefer the plain text.
> 
> Here is a comparison (first is the current implementation). The final implementation has `100` instead of `10`.
> 
> ![Screenshot_20200428_190617](https://user-images.githubusercontent.com/43481553/80517130-dd514280-8984-11ea-8f4f-a456c8f4301e.png)
> 
> ![Screenshot_20200428_190647](https://user-images.githubusercontent.com/43481553/80517141-e0e4c980-8984-11ea-9b84-c2ba452ba59f.png)

Nice! I prefer the second one, the plain text option. The InfoAlert looks a bit too aggressive to my eyes.

lorenzo-cavazzi commented:
@ciyer done
I updated https://lorenzotest.dev.renku.ch

lorenzo-cavazzi commented on src/utils/Commits.js:
```diff
@@ -91,9 +91,9 @@ function CommitsView(props) {
+  const commitsView = enhancedCommits.map((commit, i) =>
     <SingleCommit
-      key={commit.id ? commit.id : commit.readableDate}
+      key={commit.id ? commit.id : `${i}-${commit.readableDate}`}
```
the same `readableDate` shouldn't appear twice. If that happens, it may be a bug in how the date is computed

lorenzo-cavazzi replied:
Using `committed_date` instead of `authored_date` seems to be the right thing to do to avoid out of order dates. It's also in line with what GitLab shows

lorenzo-cavazzi commented:
* Switched `authored_date` for `committed_date` to keep dates consistency with GitLab and avoid duplicates
* Squashed commits and added a reference to the PR to keep the history clean

vfried commented:
> Im not sure about moving the buttons to the right. I see that it would make the page more similar to the `Cards` in the Files tab, but it would be awkward to have the number and the refresh button far from the `Commits` text they qualify.

The new changes look really good, I agree with you with the buttons and the top padding thing is a small detail. I see inside stats is a similar thing so it's ok if you leave it as it is.

lorenzo-cavazzi commented:
I had to rebase, could someone please re-approve?
